api/v1alpha1/database_types.go

```diff
@@ -10,52 +10,101 @@ type DatabaseSpec struct {
 	// Number of nodes (pods) in the cluster
 	// +required
 	Nodes int32 `json:"nodes"`
+
 	// (Optional) Storage services parameter overrides
 	// Default: (not specified)
 	// +optional
 	Service DatabaseServices `json:"service,omitempty"`
+
 	// YDB Storage cluster reference
 	// +required
 	StorageClusterRef StorageRef `json:"storageClusterRef"`
+
 	// (Optional) Name of the root storage domain
 	// Default: root
 	// +kubebuilder:validation:Pattern:=[a-zA-Z0-9]([-_a-zA-Z0-9]*[a-zA-Z0-9])?
 	// +kubebuilder:validation:MaxLength:=63
 	// +kubebuilder:default:="root"
 	// +optional
 	Domain string `json:"domain"`
-	// (Optional) Database container resource limits. Any container limits
-	// can be specified.
-	// Default: (not specified)
+
+	// (Optional) Database storage and compute resources
+	// +optional
+	Resources *DatabaseResources `json:"resources,omitempty"`  // TODO: Add validation webhook: some resources must be specified
+
+	// (Optional) Shared resources can be used by serverless databases.
 	// +optional
-	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
+	SharedResources *DatabaseResources `json:"sharedResources,omitempty"`
+
+	// (Optional) If specified, created database will be "serverless".
+	// +optional
+	ServerlessResources *ServerlessDatabaseResources `json:"serverlessResources,omitempty"`
+
 	// (Optional) Public host to advertise on discovery requests
 	// Default: ""
 	// +optional
 	PublicHost string `json:"publicHost,omitempty"`
+
 	// (Optional) YDBVersion sets the explicit version of the YDB image
 	// Default: ""
 	// +optional
 	YDBVersion string `json:"version,omitempty"`
+
 	// (Optional) Yandex Database Image
 	// +optional
 	Image PodImage `json:"image,omitempty"`
+
 	// NodeSelector is a selector which must be true for the pod to fit on a node.
 	// Selector which must match a node's labels for the pod to be scheduled on that node.
 	// More info: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
 	// +optional
 	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
 	// (Optional) If specified, the pod's scheduling constraints
 	// +optional
 	Affinity *corev1.Affinity `json:"affinity,omitempty"`
+
 	// (Optional) If specified, the pod's tolerations.
 	// +optional
 	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+
 	// (Optional) Additional custom resource labels that are added to all resources
 	// +optional
 	AdditionalLabels map[string]string `json:"additionalLabels,omitempty"`
 }
 
+type DatabaseResources struct {
+	// (Optional) Database container resource limits. Any container limits
+	// can be specified.
+	// Default: (not specified)
+	// +optional
+	ContainerResources corev1.ResourceRequirements `json:"containerResources,omitempty"`
```

Comment:
Could you elaborate here? Why do we need to set limits two times, for `Resources` and `SharedResources` properties?

Reply:
Are those k8s container limits, or some YDB specific properties?

Reply:
I replicated cms CreateDatabaseRequest here. If you want to create dedicated database you should just fill Resources propery. If you want to create shared database you should fill SharedResources instead. If you want to create serverless database you should fill ServerlessResources.

Reply:
Yes, those are container limits. Such placement was made because different database types require different resources respectively. And these differences may grow.

When you create serverless database you use container resources from shared database for it.

Encapsulating resource requirements into dedicated structures is better than collecting and validating different resource requirements based on some fields like DatabaseType. Current solution makes validation logic and general understanding of substructure much simpler.

Reply:
Also, Resources field is straightforward for general user and is currently nicely built into the CRD. [Shared|Serverless]Resources are advanced scenarios.